pkg/retryablehttp/client.go

```diff
@@ -0,0 +1,107 @@
+package retryablehttp
+
+import (
+	"io"
+	"math"
+	"math/rand"
+	"net/http"
+	"time"
+)
+
+const MaxRetryAttempts = 3
+const MinimumDelay = 500
+const MinimumDelayDuration = 500 * time.Millisecond
+const MaximumDelayDuration = 5 * time.Second
+const RandomizationFactor = 0.5
+const BackoffMultiplier = 1.5
+
+type HttpClient struct {
+	http.Client
+}
+
+func (client *HttpClient) Do(req *http.Request) (*http.Response, error) {
+	var res *http.Response
+	var err error
+	for retry := 0; ; {
+		// Reset the request body for each retry
+		if req.Body != nil {
+			body, err := req.GetBody()
+			if err != nil {
+				client.CloseIdleConnections()
+				return res, err
+			}
+			if c, ok := body.(io.ReadCloser); ok {
+				req.Body = c
+			} else {
+				req.Body = io.NopCloser(body)
+			}
+		}
+
+		res, err = client.Client.Do(req)
+		if err != nil {
+			break
+		}
+
+		shouldRetry := client.shouldRetry(req, res, err, retry)
+
+		if !shouldRetry {
+			break
+		}
+
+		sleepTime := client.sleepTime(retry)
+		retry++
+
+		time.Sleep(sleepTime)
```

Comment:
Should this be `select`ing on the `Context` from the request in order to not block the caller while it waits to retry?

An example of the concurrency pattern I'm thinking of (just pulled randomly from search, not proposing we use this package):

https://github.com/sethvargo/go-retry/blob/9bf6450b891a2dfa8064b27d3b876fcdb83be28e/retry.go#L86-L93

Reply:
Won't that effectively make all of our FGA methods asynchronous? We could go down that route, but that would mean we would need to return a done channel from all FGA methods that the caller (customer application) could listen to for the eventual result. By default, the Go standard library's [`http.Server`](https://pkg.go.dev/net/http#Server) runs every request on a separate goroutine (thread), which means that we'll only be blocking the customer application's goroutine handling that particular request. I believe this is the desired/optimal behavior. In case customers would like to run our SDK methods asynchronously, they can choose to run them in a separate goroutine if they'd like.
```go
go func() {
   res, err := fga.WriteWarrants(...)
   if err != nil {
      ...
   }
   ...
}()
```

Also note that [`time.Sleep` still yields to other goroutines](https://stackoverflow.com/a/34668153), so that's good.

Reply:
> Won't that effectively make all of our FGA methods asynchronous?

No, or at least AFAICT it shouldn't. Looking at that example I previously linked, the result of the operation wrapped in the try logic is still provided to the caller as a normal return value; not by sending it through a channel.

https://github.com/sethvargo/go-retry/blob/9bf6450b891a2dfa8064b27d3b876fcdb83be28e/retry.go#L65

```go
v, err := f(ctx)
if err == nil {
	return v, nil
}
```

Reply:
Oh, I just reread my wording:

> in order to not block the caller while it waits to retry?

Yeah, that totally sounds like I was suggesting it should be asynchronous. My bad.

I was more thinking about how the goroutine might be "blocking" say a server or background job, from shutting down/cancelling, because the method was sleeping.

Reply:
Oh, it looks like the example you linked to is using the `select` statement to check for `ctx.Done()` to support proper request cancellation (unrelated to blocking the current thread). We should probably do this as well, in which case we'd want to accept the current context as the first argument to the `Do` method @stanleyphu.

Reply:
> Oh, it looks like the example you linked to is using the select statement to check for ctx.Done() to support proper request cancellation (unrelated to blocking the current thread).

Yes, exactly! I know of the pattern (since the rest of the SDK was written with it in mind), but was grasping for the right Go terminology. Thanks for clarifying that.

Reply:
Updated the `Do` method to create a timer which then allows us to do a `select` to allow the timer to complete or the context to be cancelled.